### ISSUE TITLE
[Bug]: Better support for static analysis

### ISSUE THREAD

**Issue by DanielGSoftware** [rappasoft/laravel-livewire-tables#1741]
### What happened?

Hello there! 
To start off, I want to say that this "bug" isn't that big of a deal. I believe it to be more relevant when using static analysis tools such as PHPStan and Psalm, which is exactly what we are now implementing in our project. So, I guess this bug report is kind of a discussion to see if it's something that is wanted to be "fixed", and if so, I have a proposed solution.

Alright, let me explain.
Assume we want to use a column that extends the normal `Column`... let's say a `ComponentColumn`.
I might implement this in a table like so:
```php
ComponentColumn::make('Status', 'status')
    ->slot(fn ($foo) => //...),
```
All goes well. Let's say we also want to use attributes.
```php
ComponentColumn::make('Status', 'status')
    ->component('badges.badge')

    // Let's set attributes
    ->attributes()
    ->slot(fn ($foo) => //...),
```
When running PHPStan on level 1 or higher, it raises the following error.
`Call to an undefined method Rappasoft\LaravelLivewireTables\Views\Column::slot().`
When visiting the page this table would be present on, there is no error.

So what's going on here? It seems like we are calling `slot()` on... the Column class? But `slot()` exists in the `ComponentColumnConfiguration` trait, which is inherited by our `ComponentColumn`. The problem here is that the `attributes()` method, and many others, return `self`. This means that tools like static analysis see this as returning the class they were originally declared on (`Component`), and not any that extend this class (`ComponentColumn`). 
So
```php
public function attributes(Closure $callback): self
{
    $this->attributesCallback = $callback;

    return $this;
}
```
Should return static
```php
public function attributes(Closure $callback): static
{
    $this->attributesCallback = $callback;

    return $this;
}
```

I think that's all! 
Now, I would understand that this issue doesn't have a high priority, but I believe we are not the only ones using PHPStan and fixing this would be pretty neat. Let me know your thoughts, and if wanted, I'm pretty sure I can create a PR for this.

### Package Version

3.2.5

### PHP Version

8.3.x

### Laravel Version

11.9.2

### Theme

Tailwind 3.x

**Comment by lrljoe** [comment]
Firstly - thank you for asking the question!  Always more than happy to take PRs, as is Anthony (rappasoft)

Unless it's a break-fix, we do require PHPStan Level 5 or higher to pass for v3.0+, and when I'm running PHPStan on Level 5, I'm not seeing that error.  

We do have a couple of bits for skipping generic errors/known issues (purely so that we can focus on anything NEW that gets introduced)

Example from a new direct pull from the Master branch:
```
php vendor/bin/phpstan --level=5
Note: Using configuration file master/phpstan.neon.
⚠️  You're using a deprecated config option checkMissingIterableValueType ⚠️️

It's strongly recommended to remove it from your configuration file
and add the missing array typehints.

If you want to continue ignoring missing typehints from arrays,
add missingType.iterableValue error identifier to your ignoreErrors:

parameters:
        ignoreErrors:
                -
                        identifier: missingType.iterableValue

⚠️  You're using a deprecated config option checkGenericClassInNonGenericObjectType ⚠️️

It's strongly recommended to remove it from your configuration file
and add the missing generic typehints.

If you want to continue ignoring missing typehints from generics,
add missingType.generics error identifier to your ignoreErrors:

parameters:
        ignoreErrors:
                -
                        identifier: missingType.generics

 144/144 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%


                                                                                                                        
 [OK] No errors                                                                                                         
```

**Comment by DanielGSoftware** [comment]
Hey Joe! 
Sorry for the little delay, I didn't get any notification of your reply, I'm gonna need to have to figure out why 😛 
Anyway, it's correct that the PHPStan build of this package succeeds without any problem.
That's because in source code, there is no instance similar to the one I described above.
To verify that this error still happens if the code does exist, we can use the existing PetsTable and add the following:
```php
public function columns(): array
{
    return [
        ComponentColumn::make('test')
          ->component('livewire-tables::components.test-component')
          ->attributes(function () {})
          ->slot(fn ($row) => 'test'),
    ];
```
and then we can run PHPStan on this specific file (by default files in tests dir are not included):
`vendor/bin/phpstan analyse tests/Http/Livewire/PetsTable.php`
Result:
<img width="706" alt="Screenshot 2024-06-27 at 14 09 12" src="https://github.com/rappasoft/laravel-livewire-tables/assets/49187378/fde87a54-045f-4ec1-81f7-ac75a79eb488">

**Comment by lrljoe** [comment]
Ah!  Now I see what you're saying, thanks for the example.

It's almost certainly from the pre PHP 8.x days.

I'll update the return types, probably in the next week or so.  I'd say feel free to open a PR, but it'll be so many line changes to review that it'll probably take me as long to review a PR as to do it

**Comment by lrljoe** [comment]
Cancel my last.

So looking at it with fresh eyes, it's just the case that some Columns have additional features (e.g. "slot()"), but "Column::class" isn't aware of those.

So I think the easiest route will be:

1. Validate that none of the Child Column classes, and their traits (e.g. ComponentColumn, ComponentColumnConfiguration) use the same "extra" methods but in a different way (pretty sure that's the case).

I think it'll only be the Configuration traits that'll need checking, as the helpers are just for the frontend.  This is the bit that'll take a touch of time!

If that's the case, then:

2.  Add abstract functions for any "extra" methods into ColumnConfiguration::class and ColumnHelper::class as appropriate.

e.g, adding this into Column::class
``` 
abstract public function slot(callable $callback);
```

If you give that one a go, does it remediate the issue in this particular use case, if so, I'll look at getting it scheduled in.

**Comment by lrljoe** [comment]
I've done a quick test with the attributes() method in the HasAttributes trait:

```
    public function attributes(Closure $callback): static
    {
        $this->attributesCallback = $callback;

        return $this;
    }
```
This then seemed to work fine with:
```
            ComponentColumn::make('test')
            ->slot(fn ($row) => 'test')
            ->attributes(function () {})
            ->component('livewire-tables::components.test-component'),
```
With attributes/slot/component in any order

**Comment by DanielGSoftware** [comment]
Awesome, if you can handle changing the return types, I will make a PR for adding more complex scenario's to trigger more static analysis errors. I think the boundary of which errors I want to trigger should be related to the return types self -> static. Of course the static analysis build should fail, and then once your PR has been merged, the errors should resolve itself after a merge with dev. Is that the way to go? Or maybe it might be better to build on your PR to instantly check if the errors are gone? 🤔

**Comment by DanielGSoftware** [comment]
> I did ultimately want it passing at level 9, but quickly realised it was going to take a **lot** of PHPDoc tags to get there.

Oof, I don't want to call myself an expert on PHPStan, but to my understanding, I believe level 9 is pretty hard to get. Also I think it would be.. let's say... a more pleasurable working experience to go 1 level up per PR instead of going straight to 9. At least that's what we are doing for our project. But to be fair, we also have a lot more code 😄

**Comment by lrljoe** [comment]
Yeah, Level 9 is pretty hard to get, and basically requires you to abandon "mixed" totally, I've had a few projects require passing at level 9, which upset some devs who "hadn't yet decided" on what exactly they were passing into different elements.

In terms of including tests directory in PHPStan, typically not the done thing, as those aren't actually present in production instances, and it's hard enough to get people writing tests for their contributions without them having to also comply with that!    
Would definitely agree that it's necessary in a corporate environment tho!

In case you're interested in the present test coverage in terms of lines of code covered:
[https://app.codecov.io/gh/rappasoft/laravel-livewire-tables/tree/master/](https://app.codecov.io/gh/rappasoft/laravel-livewire-tables/tree/master/)

There is of course lots more to do, I've recently done some testing with some custom Dusk approaches thanks to the Livewire core guys sharing some useful code.

**Comment by lrljoe** [comment]
I've merged that fix into "development" branch, so it'll go into master sometime this week, and out as a release by next weekend.

**Comment by lrljoe** [comment]
Just to note, I'm also merging in a quick fix to use larastan/larastan rather than the original.  I'm also fixing one issue with the phpstan.neon file, and cleaning up one test.

**Comment by DanielGSoftware** [comment]
Awesome, I can confirm it fixed our errors! Also I wanted to thank you for the extra info about PHPStan and some other stuff, it's always helpful to hear other's stories & reasoning. I can very much see why it would indeed be "better" to not include tests for PHPStan 😄.

To continue, now that including tests is not an option, I don't know how we want to test that verify those PHPStan errors, since I was relying on using the PetsTable or adding a whole new table itself 😬. Would you have any alternative idea's?

I've shoot you a message on discord as well!

**Comment by lrljoe** [comment]
Now released as v3.2.8, please advise if anything not as expected!

**Comment by lrljoe** [comment]
@DanielGSoftware  - just to note that some missing return types were present in the latest release, I've resolved, and the project now require PHPStan Level 6 to pass prior to merging.
